# Analysis Challenge Assignment 1
## Author: Zecheng Chang

### 1. Criteria
According to Alex, his ideal schools should equiped with:
1. Safety (Low crime)
2. Urban
3. Diversity
4. Quality

Based on these requirment, I'll look into those data I have and give him some professional advices.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

###  Import, clean, prepare, and merge data¶

In [2]:
cs = pd.read_csv('ACA1_dataset/CollegeScorecard.csv')

/Users/czc/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (7,9,1427,1542,1561,1575,1725,1726,1727,1728) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
cs_dict = pd.read_csv('ACA1_dataset/CollegeScorecardDataDictionary-09-12-2015.csv')

In [4]:
cs.head(5)

UNITID    OPEID  opeid6                               INSTNM        CITY  \
0  100654   100200    1002             Alabama A & M University      Normal   
1  100663   105200    1052  University of Alabama at Birmingham  Birmingham   
2  100690  2503400   25034                   Amridge University  Montgomery   
3  100706   105500    1055  University of Alabama in Huntsville  Huntsville   
4  100724   100500    1005             Alabama State University  Montgomery   

  STABBR         ZIP                                       AccredAgency  \
0     AL       35762  Southern Association of Colleges and Schools C...   
1     AL  35294-0110  Southern Association of Colleges and Schools C...   
2     AL  36117-3553  Southern Association of Colleges and Schools C...   
3     AL       35899  Southern Association of Colleges and Schools C...   
4     AL  36104-0271  Southern Association of Colleges and Schools C...   

                          INSTURL  \
0                   www.aamu.edu/   
1                     www.uab.edu   
2       www.amridgeuniversity.edu   
3                     www.uah.edu   
4  www.alasu.edu/email/index.aspx   

                                              NPCURL  ...  \
0     galileo.aamu.edu/netpricecalculator/npcalc.htm  ...   
1    www.collegeportraits.org/AL/UAB/estimator/agree  ...   
2  tcc.noellevitz.com/(S(miwoihs5stz5cpyifh4nczu0...  ...   
3                                    finaid.uah.edu/  ...   
4  www.alasu.edu/cost-aid/forms/calculator/index....  ...   

   PELL_RPY_3YR_RT_SUPP  NOPELL_RPY_3YR_RT_SUPP  FEMALE_RPY_3YR_RT_SUPP  \
0        0.443586698337           0.45251396648          0.484818805093   
1        0.732297852901          0.795454545455          0.760521885522   
2        0.600877192982                   0.775          0.610465116279   
3        0.738341968912           0.85140562249          0.768378650554   
4        0.320016820858          0.380829015544          0.364423717522   

   MALE_RPY_3YR_RT_SUPP  FIRSTGEN_RPY_3YR_RT_SUPP  \
0                 0.406            0.423580786026   
1        0.755344418052            0.751552795031   
2        0.686567164179            0.632653061224   
3        0.798440979955            0.778156996587   
4        0.286530612245            0.315737051793   

   NOTFIRSTGEN_RPY_3YR_RT_SUPP  C150_L4_POOLED_SUPP  C150_4_POOLED_SUPP  \
0               0.442452830189                  NaN    0.30871831265508   
1               0.765074770863                  NaN    0.50854980789381   
2                         0.65                  NaN   PrivacySuppressed   
3               0.791666666667                  NaN    0.47821134615384   
4               0.334633385335                  NaN    0.25748195545488   

   C200_L4_POOLED_SUPP  C200_4_POOLED_SUPP  
0                  NaN    0.35168556430446  
1                  NaN     0.4730923834537  
2                  NaN   PrivacySuppressed  
3                  NaN    0.49412939796716  
4                  NaN    0.29365162671232  

[5 rows x 1729 columns]

In [5]:
cs_dict.head(3)

NAME OF DATA ELEMENT Year dev-category developer-friendly name  \
0         Unit ID for institution  NaN         root                      id   
1  8-digit OPE ID for institution  NaN         root                 ope8_id   
2  6-digit OPE ID for institution  NaN         root                 ope6_id   

  VARIABLE NAME API data type label VALUE LABEL SCORECARD? Y/N SOURCE NOTES  
0        UNITID       integer   NaN   NaN   NaN            NaN  IPEDS   NaN  
1         OPEID       integer   NaN   NaN   NaN            NaN  IPEDS   NaN  
2        opeid6       integer   NaN   NaN   NaN            NaN  IPEDS   NaN

In [6]:
cs.shape

(7804, 1729)

In [7]:
cs_dict.shape

(1953, 12)

College Scorecard Data Dict has more features than College Scorecard Data

In [8]:
cs_dict['VARIABLE NAME'].isnull().sum()

233

There are 233 variable name is NaN in dictionary

In [9]:
# find the unique variable names in dictionary
vb = list(cs_dict['VARIABLE NAME'].unique())

In [10]:
# find out if there are features in dataset than in dataset dictionary
for i in cs.columns:
    if i not in vb:
        print(i)

C200_4
C200_L4
D200_4
D200_L4
C200_4_POOLED
C200_L4_POOLED
poolyrs200
D200_4_POOLED
D200_L4_POOLED


These are the columns in College Scorecard Dataset but not in the dictionary

In [105]:
crime = pd.read_csv('ACA1_dataset/Crime_2015.csv',thousands=',')
# there are commas in Crime, set thousands = ',' can omit them.

In [102]:
crime.head()

MSA  ViolentCrime  Murder  Rape  Robbery  \
0      Abilene, TX M.S.A.         412.5     5.3  56.0     78.4   
1        Akron, OH M.S.A.         238.4     5.1  38.2     75.2   
2       Albany, GA M.S.A.         667.9     7.8  30.4    157.9   
3       Albany, OR M.S.A.         114.3     2.5  28.2     20.7   
4  Albuquerque, NM M.S.A.         792.6     6.1  63.8    206.7   

   AggravatedAssault  PropertyCrime  Burglary   Theft  MotorVehicleTheft  \
0              272.8         3609.0     852.0  2493.6              263.4   
1              119.8         2552.4     575.3  1853.0              124.1   
2              471.8         3894.1    1099.6  2652.8              141.7   
3               63.0         3208.4     484.6  2476.1              247.7   
4              516.0         4607.8     883.4  3047.6              676.9   

  State         City  
0    TX      Abilene  
1    OH        Akron  
2    GA       Albany  
3    OR       Albany  
4    NM  Albuquerque

### Done with importing datasets, now, it's time to explore more about these datasets

### 1. Urban

Since urbanization info is in the `CollegeScorecard` dataset and it has more features than crime dataset, let's explore it first.

According to the dictionary, `LOCALE` & `locale2` are the variables representing urbanization of the school, however, there is no description about this 2 variables in documentation. I need to figure out what these 2 variable exactly mean.

In [20]:
cs['locale2'].describe()

count    0.0
mean     NaN
std      NaN
min      NaN
25%      NaN
50%      NaN
75%      NaN
max      NaN
Name: locale2, dtype: float64

In [16]:
cs['locale2'].isnull().sum()

7804

Since there is only Nan in `locale2`, it can be ignored.

In [19]:
cs['LOCALE'].describe()

count    7380.000000
mean       19.589024
std         9.380431
min        11.000000
25%        12.000000
50%        21.000000
75%        22.000000
max        43.000000
Name: LOCALE, dtype: float64

In [28]:
cs[(cs['CITY'] == "New York") & (cs['STABBR'] == 'NY')]['LOCALE']
# make sure it's New York of New York State, since there are 7 city named New York in US.

692     11.0
2381    11.0
2388    11.0
2389    11.0
2390    11.0
        ... 
7253    11.0
7327    11.0
7344    11.0
7582     NaN
7733     NaN
Name: LOCALE, Length: 92, dtype: float64

New York has the smallest number of `LOCALE`, I guess the smaller the `LOCALE`, the more urban the city.
We need a another city to verify this

In [25]:
cs[cs['LOCALE'] == 43][['CITY','STABBR']]

CITY STABBR
30        Marion     AL
32    Livingston     AL
34        Marion     AL
65        Seward     AK
67        Valdez     AK
...          ...    ...
5340    Mahnomen     MN
5556     Houston     MO
5628       Sells     AZ
6099     Oakwood     VA
6835      Baraga     MI

[62 rows x 2 columns]

These are cities with only a population of thousands people and kind of out of nowher with little urbanization. This helps verfifies my guess.  
**I will include `LOCALE` in the later dataset for urbanization**

### 2. Diversity

According to documentation, there are some variables represent the race of undergraduate degree-seeking student. This is a good feature that can be used to quantify the diveristy of school.

In [37]:
race_info = [i for i in cs.columns if 'UGDS_' in i]

In [38]:
race_info

['UGDS_WHITE',
 'UGDS_BLACK',
 'UGDS_HISP',
 'UGDS_ASIAN',
 'UGDS_AIAN',
 'UGDS_NHPI',
 'UGDS_2MOR',
 'UGDS_NRA',
 'UGDS_UNKN',
 'UGDS_WHITENH',
 'UGDS_BLACKNH',
 'UGDS_API',
 'UGDS_AIANOld',
 'UGDS_HISPOld']

`USDS_UNKN` is student whose race is unknow, which can be ignored

In [39]:
race_info.remove('UGDS_UNKN')

I'll use **Simpson’s Index** to quantify the diversity later

![](Simpsons_Index.png)

### 3. Quality

Since Alex decided to pursue a degree in Education and he wanted to go to a school with hige quality, I'll be using 3 variables to indicate the quality of shcool, `SAT_AVG` and `PFTFAC`  
`SAT_AVG` is the average sat score of students admitted, since higher score students must apply for a better school.  
`PFTFAC` is the proportion of full-time faculty, the more full-time faculties, the higher the quality of school.

I was thinking of using median earnings of students to be the indicator or quality as well, but there are only NaN values.

So the independent variabels mentioned above will be included in the final dataset.

### Slicing the columns which we need in `CollegeScored`.

In order to merge with `crime` data later, keep `CITY` and `STABBR` as well
Since Alex wants to get a degree in Education, I'll only include school have awared degree in Education. `PCIP13` gives this info

In [122]:
df = cs.loc[:,['INSTNM','CITY','STABBR','LOCALE','SAT_AVG','PFTFAC','PCIP13']]

In [123]:
df_temp = cs.loc[:,race_info]

In [124]:
df = pd.concat([df,df_temp],axis=1)

In [125]:
df.head()

INSTNM        CITY STABBR  LOCALE  SAT_AVG  \
0             Alabama A & M University      Normal     AL    12.0    823.0   
1  University of Alabama at Birmingham  Birmingham     AL    12.0   1146.0   
2                   Amridge University  Montgomery     AL    12.0      NaN   
3  University of Alabama in Huntsville  Huntsville     AL    12.0   1180.0   
4             Alabama State University  Montgomery     AL    12.0    830.0   

   PFTFAC  PCIP13  UGDS_WHITE  UGDS_BLACK  UGDS_HISP  UGDS_ASIAN  UGDS_AIAN  \
0  0.8856  0.1490      0.0279      0.9501     0.0089      0.0022     0.0012   
1  0.9106  0.0862      0.5987      0.2590     0.0258      0.0518     0.0026   
2  0.6721  0.0000      0.2919      0.4224     0.0093      0.0031     0.0031   
3  0.6555  0.0173      0.7012      0.1310     0.0338      0.0364     0.0145   
4  0.6641  0.2150      0.0161      0.9285     0.0114      0.0015     0.0009   

   UGDS_NHPI  UGDS_2MOR  UGDS_NRA  UGDS_WHITENH  UGDS_BLACKNH  UGDS_API  \
0     0.0010     0.0000    0.0002           NaN           NaN       NaN   
1     0.0007     0.0344    0.0140           NaN           NaN       NaN   
2     0.0031     0.0000    0.0000           NaN           NaN       NaN   
3     0.0002     0.0161    0.0329           NaN           NaN       NaN   
4     0.0007     0.0064    0.0207           NaN           NaN       NaN   

   UGDS_AIANOld  UGDS_HISPOld  
0           NaN           NaN  
1           NaN           NaN  
2           NaN           NaN  
3           NaN           NaN  
4           NaN           NaN

### 4. Crime

In [126]:
crime.head(3)

MSA  ViolentCrime  Murder  Rape  Robbery  AggravatedAssault  \
0  Abilene, TX M.S.A.         412.5     5.3  56.0     78.4              272.8   
1    Akron, OH M.S.A.         238.4     5.1  38.2     75.2              119.8   
2   Albany, GA M.S.A.         667.9     7.8  30.4    157.9              471.8   

   PropertyCrime  Burglary   Theft  MotorVehicleTheft State     City  \
0         3609.0     852.0  2493.6              263.4    TX  Abilene   
1         2552.4     575.3  1853.0              124.1    OH    Akron   
2         3894.1    1099.6  2652.8              141.7    GA   Albany   

   total_crime  
0       8043.0  
1       5581.5  
2       9124.0

Just sum up all kinds of crimes

In [127]:
crime['total_crime'] = crime.sum(numeric_only=True,axis=1)

In [128]:
crime_new = crime.loc[:,['City','State','total_crime']]

In [129]:
crime_new.head(3)

City State  total_crime
0  Abilene    TX      16086.0
1    Akron    OH      11163.0
2   Albany    GA      18248.0

## Merge 2 data frames.

In [134]:
data = pd.merge(df,crime_new,left_on=['CITY','STABBR'], right_on=['City','State'])

In [135]:
data.head(3)

INSTNM        CITY STABBR  LOCALE  SAT_AVG  \
0  University of Alabama at Birmingham  Birmingham     AL    12.0   1146.0   
1          Birmingham Southern College  Birmingham     AL    12.0   1177.0   
2        Herzing University-Birmingham  Birmingham     AL    21.0      NaN   

   PFTFAC  PCIP13  UGDS_WHITE  UGDS_BLACK  UGDS_HISP  ...  UGDS_2MOR  \
0  0.9106  0.0862      0.5987      0.2590     0.0258  ...     0.0344   
1  0.6479  0.0731      0.8103      0.0940     0.0305  ...     0.0076   
2  0.3077  0.0000      0.3491      0.5881     0.0063  ...     0.0314   

   UGDS_NRA  UGDS_WHITENH  UGDS_BLACKNH  UGDS_API  UGDS_AIANOld  UGDS_HISPOld  \
0     0.014           NaN           NaN       NaN           NaN           NaN   
1     0.000           NaN           NaN       NaN           NaN           NaN   
2     0.000           NaN           NaN       NaN           NaN           NaN   

         City  State  total_crime  
0  Birmingham     AL      16107.8  
1  Birmingham     AL      16107.8  
2  Birmingham     AL      16107.8  

[3 rows x 23 columns]

### Continue cleaning

In [136]:
# exclue schools haven't awared any degree in Education yet.
data = data[data['PCIP13'] != 0]

In [137]:
data['PCIP13'].describe()

count    718.000000
mean       0.077965
std        0.110139
min        0.000200
25%        0.020300
50%        0.050150
75%        0.096350
max        1.000000
Name: PCIP13, dtype: float64

### calculate Simpsion's Index

In [146]:
race_info

['UGDS_WHITE',
 'UGDS_BLACK',
 'UGDS_HISP',
 'UGDS_ASIAN',
 'UGDS_AIAN',
 'UGDS_NHPI',
 'UGDS_2MOR',
 'UGDS_NRA',
 'UGDS_WHITENH',
 'UGDS_BLACKNH',
 'UGDS_API',
 'UGDS_AIANOld',
 'UGDS_HISPOld']

In [155]:
data[race_info].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1084 entries, 0 to 3306
Data columns (total 13 columns):
UGDS_WHITE      723 non-null float64
UGDS_BLACK      723 non-null float64
UGDS_HISP       723 non-null float64
UGDS_ASIAN      723 non-null float64
UGDS_AIAN       723 non-null float64
UGDS_NHPI       723 non-null float64
UGDS_2MOR       723 non-null float64
UGDS_NRA        723 non-null float64
UGDS_WHITENH    0 non-null float64
UGDS_BLACKNH    0 non-null float64
UGDS_API        0 non-null float64
UGDS_AIANOld    0 non-null float64
UGDS_HISPOld    0 non-null float64
dtypes: float64(13)
memory usage: 118.6 KB


Exlcude UGDS_WHITENH, UGDS_BLACKNH, UGDS_API, UGDS_AIANOld, UGDS_HISPOld since they only have NaN

In [156]:
data.drop(columns=['UGDS_WHITENH', 'UGDS_BLACKNH', 'UGDS_API', 'UGDS_AIANOld', 'UGDS_HISPOld'], inplace=True)

In [162]:
# update race info
race_info = [i  for i in data.columns if 'UGDS_' in i]

In [167]:
# function to calculate Simpsion index
def simpsion_index(races):
    num_col = len(race_info)
    sum_sqr = 0
    for i in race_info:
        sum_sqr += races[i]**2
    return sum_sqr
    

In [168]:
data['SI'] = data.apply(simpsion_index,axis=1)

In [169]:
data.head(3)

INSTNM        CITY STABBR  LOCALE  SAT_AVG  \
0  University of Alabama at Birmingham  Birmingham     AL    12.0   1146.0   
1          Birmingham Southern College  Birmingham     AL    12.0   1177.0   
5                   Samford University  Birmingham     AL    21.0   1153.0   

   PFTFAC  PCIP13  UGDS_WHITE  UGDS_BLACK  UGDS_HISP  UGDS_ASIAN  UGDS_AIAN  \
0  0.9106  0.0862      0.5987      0.2590     0.0258      0.0518     0.0026   
1  0.6479  0.0731      0.8103      0.0940     0.0305      0.0390     0.0102   
5  0.9939  0.0784      0.8114      0.0664     0.0664      0.0080     0.0030   

   UGDS_NHPI  UGDS_2MOR  UGDS_NRA        City State  total_crime        SI  
0     0.0007     0.0344    0.0140  Birmingham    AL      16107.8  0.430258  
1     0.0000     0.0076    0.0000  Birmingham    AL      16107.8  0.668035  
5     0.0007     0.0100    0.0304  Birmingham    AL      16107.8  0.668286

In [171]:
# update data dataframe again

data = data.loc[:,['INSTNM','CITY','STABBR','LOCALE','SAT_AVG','PFTFAC','PCIP13','total_crime','SI']]

In [172]:
data.head(3)

INSTNM        CITY STABBR  LOCALE  SAT_AVG  \
0  University of Alabama at Birmingham  Birmingham     AL    12.0   1146.0   
1          Birmingham Southern College  Birmingham     AL    12.0   1177.0   
5                   Samford University  Birmingham     AL    21.0   1153.0   

   PFTFAC  PCIP13  total_crime        SI  
0  0.9106  0.0862      16107.8  0.430258  
1  0.6479  0.0731      16107.8  0.668035  
5  0.9939  0.0784      16107.8  0.668286

### Normalize the numbers since the variables are not in the same scale

In [177]:
# drop schools with missing values
data.dropna(axis=0,inplace=True)

In [182]:
data_final = data.copy()

In [183]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 406 entries, 0 to 3269
Data columns (total 9 columns):
INSTNM         406 non-null object
CITY           406 non-null object
STABBR         406 non-null object
LOCALE         406 non-null float64
SAT_AVG        406 non-null float64
PFTFAC         406 non-null float64
PCIP13         406 non-null float64
total_crime    406 non-null float64
SI             406 non-null float64
dtypes: float64(6), object(3)
memory usage: 31.7+ KB


In [209]:
for i in ['LOCALE','SAT_AVG','PFTFAC','PCIP13','total_crime','SI']:
    data_final[i] = (data[i] - data[i].mean())/(data[i].std())
    

In [213]:
data_final

INSTNM        CITY STABBR    LOCALE  \
0     University of Alabama at Birmingham  Birmingham     AL -0.232833   
1             Birmingham Southern College  Birmingham     AL -0.232833   
5                      Samford University  Birmingham     AL  1.976966   
14               Alabama State University  Montgomery     AL -0.232833   
15        Auburn University at Montgomery  Montgomery     AL -0.232833   
...                                   ...         ...    ...       ...   
3246    University of Wisconsin-Green Bay   Green Bay     WI -0.232833   
3260   University of Wisconsin-Eau Claire  Eau Claire     WI  0.012700   
3266                   Viterbo University   La Crosse     WI  0.012700   
3268    University of Wisconsin-La Crosse   La Crosse     WI  0.012700   
3269      University of Wisconsin-Oshkosh     Oshkosh     WI  0.012700   

       SAT_AVG    PFTFAC    PCIP13  total_crime        SI  final_score  
0     0.779039  0.883059  0.067725     1.120205 -0.507568    -1.029558  
1     1.044692 -0.290619 -0.085997     1.120205  0.940656    -0.035346  
5     0.839025  1.255222 -0.023804     1.120205  0.942181     2.846066  
14   -1.928907 -0.218242  1.579130     0.884978  2.127970    -0.063416  
15   -0.729184  0.952756  0.706082     0.884978 -0.689007    -1.695033  
...        ...       ...       ...          ...       ...          ...  
3246  0.127761  1.255222 -0.225638    -1.685286  1.470447     3.614391  
3260  0.470539  1.139954 -0.037886    -1.510917  1.801045     4.129908  
3266  0.299150 -1.342328 -0.073089    -0.922819  1.912918     2.326848  
3268  0.641928  1.203396  0.156907    -0.922819  1.754719     3.612900  
3269 -0.215017  1.229756  0.390424    -1.436305  1.597555     3.553929  

[406 rows x 10 columns]

### The smaller the number of `LOCALE` & 'total_crime' & 'SI', the better the school, so using inverse of these variables

In [219]:
data_final['final_score'] = -1*data_final['LOCALE'] + \
                            0.5*(data_final['SAT_AVG'] + data_final['PFTFAC']) + \
                            -1*data_final['total_crime'] + \
                            -1*data_final['SI']
### Each characteristic of school are equally important, but there are 2 variables(SAT_AVG&PFTFAC) represent the quality of school, so use the mean of them.

In [221]:
data_final.sort_values(by='final_score',ascending=False).head(20)

INSTNM        CITY STABBR    LOCALE  \
2684                  University of St Thomas     Houston     TX -0.478366   
2511                         Brown University  Providence     RI -0.232833   
2675               Houston Baptist University     Houston     TX -0.478366   
585   Metropolitan State University of Denver      Denver     CO -0.478366   
1983        Southern New Hampshire University  Manchester     NH -0.232833   
2043                  University of Rochester   Rochester     NY -0.232833   
1780              University of Detroit Mercy     Detroit     MI -0.478366   
1701                 Johns Hopkins University   Baltimore     MD -0.478366   
1773         University of Michigan-Ann Arbor   Ann Arbor     MI -0.232833   
587                          Regis University      Denver     CO -0.478366   
1403                    Vanderbilt University   Nashville     TN -0.478366   
666                       American University  Washington     DC -0.478366   
2415                        Carlow University  Pittsburgh     PA -0.478366   
1192                    North Park University     Chicago     IL -0.478366   
3070                       Liberty University   Lynchburg     VA  0.012700   
667                       American University  Washington     DC -0.478366   
401            Point Loma Nazarene University   San Diego     CA -0.478366   
1193                    North Park University     Chicago     IL -0.478366   
1156        University of Illinois at Chicago     Chicago     IL -0.478366   
405                San Diego State University   San Diego     CA -0.478366   

       SAT_AVG    PFTFAC    PCIP13  total_crime        SI  final_score  
2684  0.564803  1.259690 -0.054314    -2.395926 -1.497379     5.283918  
2511  3.169916  0.768237 -0.777160    -1.024170 -1.698042     4.924123  
2675  0.170608 -0.576109  0.102928    -2.395926 -1.880416     4.551958  
585  -0.703476  1.282476 -0.793588    -2.977000 -0.584344     4.329210  
1983 -0.772031  1.282476 -0.633999    -1.030604 -2.595784     4.114443  
2043  2.552915  1.071151 -0.878077    -0.814325 -1.133849     3.993040  
1780  0.744761  1.282476 -0.822925    -0.674842 -1.325894     3.492720  
1701  3.109929  0.673521 -0.929709     0.202612 -1.313601     3.481080  
1773  2.544346  0.400542 -0.781854    -1.141164 -0.586348     3.432789  
587   0.273441 -1.830206 -0.917974    -2.977000 -0.560268     3.237251  
1403  3.598388  0.550658 -0.396962     0.135497 -0.737449     3.154841  
666   1.738817 -0.610064 -0.872210    -0.973760 -1.013223     3.029726  
2415 -0.934851  1.282476  0.801132    -1.214654 -1.147775     3.014608  
1192 -0.215017  1.282476  0.024307    -0.810547 -1.167913     2.990556  
3070 -0.215017 -0.355849 -0.244413    -1.702876 -1.541682     2.946426  
667   1.738817 -0.610064 -0.872210    -0.819520 -1.013223     2.875486  
401   0.856164  1.282476 -0.860475    -0.973052 -0.336254     2.856992  
1193 -0.215017  1.282476  0.024307    -0.673602 -1.167913     2.853611  
1156  0.316289 -0.264260 -0.509613    -0.810547 -1.528105     2.843033  
405   0.299150 -0.770903 -0.678590    -0.973052 -1.596609     2.812151

## This 20 Universities are the ones I recommend to Alex based on his preference.